toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,292 @@
+
+```console
+cf create-secret SECRET_NAME SECRET_TYPE -p <CREDENTIALS_INLINE>
+```
+
+```console
+cf create-secret SECRET_NAME SECRET_TYPE -f <FILE_PATH_TO_CREDENTIALS_FILE>
+```
+
+* `-p` reads inline/literal credentials, no matter the `SECRET_TYPE`
+* `-f` reads credentials from a file
+
+This command will ultimately create a `POST` request to `/v3/secrets`. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets" \
+  -X POST \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "json",
+    "name": "my-app-secrets",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "my-secret"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    },
+    "relationships": {
+      "space": {
+        "data": {
+          "guid": "7304bc3c-7010-11ea-8840-48bf6bec2d78"
+        }
+      }
+    }
+  }'
+```
+
+### Updating a `secret`
+Users will be able to update a `secret` using similar commands:
+
+```console
+cf update-secret SECRET_NAME SECRET_TYPE -p <CREDENTIALS_INLINE>
+```
+
+```console
+cf update-secret SECRET_NAME SECRET_TYPE -f <FILE_PATH_TO_CREDENTIALS_FILE>
+```
+
+Initially apps will need to be restarted to receive `secret` updates, but eventually we may be able to support updates that don't require a restart.
+The "Update secret" endpoint in the CF API will take a `PATCH` request. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]" \
+  -X PATCH \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "json",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "updated-secret-value"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    }
+  }'
+```
+
+### Reading a `secret` and Its Value
+Users that have permission to view environment variables for an app will be able to view `secret` values within a `space`. We will introduce a separate endpoint for retrieving these (similar to how app environment variables and service instance credentials are retrieved): `GET /v3/secrets/:guid/value`.
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]/value" \
+  -X GET \
+  -H "Authorization: bearer [token]"
+  
+"value" : {
+  "some-json-key" : "some-json-value"
+}
+```
+
+To support this, Cloud Controller will need to have both read and write permissions for the `secret`'s corresponding credential in CredHub.
+
+### Creating a `secret_binding`
+To create `secret_bindings`, users will use a new CLI command: `cf bind-secret`. This will create a `POST` request to `/v3/secret_bindings`.
+
+```console
+cf bind-secret APP_NAME SECRET_NAME BINDING_NAME
+```
+
+### CLI Changes
+Work will need to be done to implement the following commands:
+- `cf create-secret`
+- `cf update-secret`
+- `cf bind-secret`
```

Comment:
**thought** For a user with no knowledge of the inner workings of this, they may create a secret, bind it and expect to see it somewhere in the apps ENV and be confused. Not sure we can do much here beyond making sure the commands behaviour are well documented.

Reply:
I think that's fine. It's a new feature that I expect primarily to be used by users coming from Kubernetes. You did remind me though that it would be interesting to map secrets to env vars like in k8s. This could be useful if the secrets are stored in CredHub. I'm not sure what a clean UX for that would look like, though.